src/wasm/wasm-type.cpp

```diff
@@ -1445,33 +1445,92 @@ void Canonicalizer::scanType(Type* type) {
 }
 
 void Canonicalizer::findSelfReferentialHeapTypes() {
-  // Calculate the fixed point of the reachability relation.
-  auto fixedPoint = children;
-  bool changed;
-  do {
-    changed = false;
-    for (auto& entry : fixedPoint) {
-      auto& succs = entry.second;
-      std::unordered_set<TypeID> newSuccs;
-      for (auto& other : succs) {
-        auto& otherSuccs = fixedPoint[other];
-        newSuccs.insert(otherSuccs.begin(), otherSuccs.end());
+  // Use Tarjan's strongly connected components algorithm on the parent-child
+  // graph to find self-referential types in O(|V|+|E|) time. All HeapTypes in a
+  // strongly connected component with multiple elements must be
+  // self-referential because it is mutually recursive with all other HeapTypes
+  // in its strongly connected component. HeapTypes in strongly connected
+  // components of size one may also be self-referential, but it is trivial to
+  // find these because they must be their own direct children. See
+  // https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm.
+
+  // Get the HeapType children of a HeapType, skipping all intermediate Types.
+  auto getChildren = [&](HeapType type) {
+    std::unordered_set<HeapType> results;
+    std::function<void(TypeID, bool)> visit = [&](TypeID id, bool isRoot) {
+      if (isRoot || typeLocations.count(Type(id))) {
+        auto it = children.find(id);
+        if (it != children.end()) {
+          for (TypeID child : it->second) {
+            visit(child, false);
+          }
+        }
+      } else {
+        results.insert(HeapType(id));
+      }
+    };
+    visit(type.getID(), true);
+    return results;
+  };
+
+  // The Tarjan's algorithm stack. Similar to a DFS stack, but elements are only
+  // popped off once they are committed to a strongly connected component.
+  // HeapTypes stay on the stack after they are visited iff they have a back
+  // edge to a HeapType earlier in the stack.
+  std::vector<HeapType> stack;
+  std::unordered_set<HeapType> stackElems;
+  // Indices assigned to each HeapType in the order they are visited.
+  std::unordered_map<HeapType, size_t> indices;
+  // The smallest index of the HeapTypes reachable from each HeapType.
+  std::unordered_map<HeapType, size_t> minReachable;
+
+  std::function<void(HeapType)> visit = [&](HeapType curr) {
+    size_t index = indices.size();
+    indices[curr] = index;
+    minReachable[curr] = index;
+    stack.push_back(curr);
+    stackElems.insert(curr);
+
+    for (HeapType child : getChildren(curr)) {
+      if (!indices.count(child)) {
+        // Child has not been visited yet; visit it.
+        visit(child);
+        minReachable[curr] = std::min(minReachable[curr], minReachable[child]);
+      } else if (stackElems.count(child)) {
+        // Child was already visited but not committed to a strongly connected
+        // component.
+        minReachable[curr] = std::min(minReachable[curr], indices[child]);
```

Comment:
This implements the original algorithm but as pointed out in the wikipedia page (the comment with "`that is deliberate and from the original paper`"), that is confusing and unnecessary. The entire code could be
```cpp
if (!indices.count(child)) {
  // Visit the child for the first time, and lower down update minReachable
  // afterwards.
  visit(child);
} else if (!stackElems.count(child)) {
  // The child is not on the stack, so skip the update of minReachable
  // afterwards.
  continue;
}
minReachable[curr] = std::min(minReachable[curr], minReachable[child]);
```
This is shorter and also properly preserves the invariant.

Alternatively, you can replace `indices[child]` with `minReachable[child]` in your code (which would make the two update lines identical, and which my version deduplicates). I don't feel strongly between the two.

(The difference is not observable in the algorithm, so both are correct, of course, as the crucial thing is that minReachable is lower than the current node, and not by how much.)

Reply:
That comment in the Wikipedia pseudocode is calling the line out because it is suprising but necessary. Later in the article it says,

> Note that v.lowlink := min(v.lowlink, w.index) is the correct way to update v.lowlink if w is on stack. Because w is on the stack already, (v, w) is a back-edge in the DFS tree and therefore w is not in the subtree of v. Because v.lowlink takes into account nodes reachable only through the nodes in the subtree of v we must stop at w and use w.index instead of w.lowlink

Reply:
To be fair, I don't quite understand why this is necessary, but I'd rather not diverge from the Wikipedia explanation without deep understanding and confidence.

Reply:
Oh, yes - I missed that. Yes, I was wrong, it would be incorrect to mark a lower index in that case. (It would still work I think, but the invariant would be wrong: `v` cannot necessarily reach everything `w` reaches, if `w` was before it on the stack.)

Sgtm to keep this part as is.